File: LATIHAN_ARRAY/js/c4.js
function manageQueue(arr, operations) {
    // clone array supaya tidak mengubah array asli
    const queue = [...arr];

    // statistik
    const stats = {
        enqueued: 0,
        dequeued: 0,
        operations: operations.length
    };

    for (let op of operations) {
        if (op.startsWith("enqueue:")) {
            const value = op.split(":")[1];
            queue.push(value); // enqueue ke akhir
            stats.enqueued++;
        } else if (op === "dequeue") {
            if (queue.length > 0) {
                queue.shift(); // dequeue dari awal
                stats.dequeued++;
            }
        }
    }

    return {
        result: queue,
        stats
    };
}

// driver code
console.log(manageQueue([1, 2], ["enqueue:3", "dequeue", "enqueue:4"]));
console.log(manageQueue([], ["enqueue:x", "enqueue:y", "dequeue"]));

// | Baris                    | Fungsi               |
// | ------------------------ | -------------------- |
// | `for...of`               | Loop setiap operasi  |
// | `startsWith("enqueue:")` | Deteksi enqueue      |
// | `split(":")[1]`          | Ambil nilai          |
// | `push()`                 | Masuk belakang queue |
// | `shift()`                | Keluar depan queue   |
// | `stats++`                | Hitung operasi       |

File: challenge/LATIHAN_ARRAY/rs/src/c_array/c4.rs
#[derive(Debug)]
pub struct QueueResult {
    pub result: Vec<String>, // hasil akhir queue
    pub stats: QueueStats,   // statistik operasi
}

#[derive(Debug)]
pub struct QueueStats {
    pub enqueued: usize,
    pub dequeued: usize,
    pub operations: usize,
}

pub fn manage_queue(arr: Vec<String>, operations: Vec<&str>) -> QueueResult {
    let mut queue = arr; // langsung gunakan arr, clone nggak perlu
    let mut enqueued = 0;
    let mut dequeued = 0;

    for op in operations.iter() {
        if op.starts_with("enqueue:") {
            let value = op[8..].to_string();
            queue.push(value);
            enqueued += 1;
        } else if op == &"dequeue" {
            if !queue.is_empty() {
                queue.remove(0); // FIFO: shift dari awal
                dequeued += 1;
            }
        }
    }

    let stats = QueueStats {
        enqueued,
        dequeued,
        operations: operations.len(),
    };

    QueueResult { result: queue, stats }
}
